File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
    bool isSafe(vector<vector<bool>> &board, int r, int c){
        int tr=r;
        int tc=c;
        //left diag
        while(tr>=0 && tc>=0){
            if(board[tr][tc])
                return false;
            tr--;
            tc--;
        }

        tr=r;
        tc=c;
        //right diag
        while(tr>=0 && tc<board.size()){
            if(board[tr][tc])
                return false;
            tr--;
            tc++;
        }

        tr=r;
        tc=c;
        //vertical
        while(tr>=0){
            if(board[tr][tc])
                return false;
            tr--;
        }
        
        return true;
    }

    void find_ans(vector<vector<string>> &ans, vector<vector<bool>> &board){
        vector<string> tv;
        for(int i=0;i<board.size();i++){
            string temp="";
            for(int j=0;j<board.size();j++){
                if(board[i][j])
                    temp+="Q";
                else
                    temp+=".";
            }
            tv.push_back(temp);
        }

        ans.push_back(tv);
    }
    void nq(vector<vector<bool>> &board, int r, vector<vector<string>> &ans){
        if(r==board.size()){
            find_ans(ans,board);
            return;
        }

        for(int c=0;c<board.size();c++){
            if(isSafe(board,r,c)){
                board[r][c]=true;
                nq(board,r+1,ans);
                board[r][c]=false;
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<bool>> board(n,vector<bool>(n,false));
        vector<vector<string>> ans;
        nq(board,0,ans);
        return ans;
    }
};

File: 0051-n-queens/0051-n-queens.py
class Solution:

    def find_ans(self,board,ans):
        temp=[]
        for i in board:
            row = ""
            for j in i:
                if j==True:
                    row+="Q"
                else:
                    row+="."
            temp.append(row)
        ans.append(temp)

    def isSafe(self,board,r,c):
        tr,tc=r,c
        while(tr>=0 and tc>=0):
            if(board[tr][tc]):
                return False
            tr-=1
            tc-=1
        
        tr,tc=r,c
        while(tc<len(board) and tr>=0):
            if board[tr][tc]:
                return False
            tr-=1
            tc+=1
        
        tr=r
        while(tr>=0):
            if board[tr][c]:
                return False
            tr-=1
        
        return True

    def nq(self,board,n,r,ans):
        if(r==n):
            self.find_ans(board,ans)
            return
        
        for c in range(n):
            if self.isSafe(board,r,c):
                board[r][c]=True
                self.nq(board,n,r+1,ans)
                board[r][c]=False

    def solveNQueens(self, n: int) -> List[List[str]]:
        board=[]
        ans=[]

        for i in range(n):
            temp = []
            for j in range(n):
                temp.append(False)
            board.append(temp)
        
        self.nq(board,n,0,ans)
        return ans